### PR TITLE
Always run golden test during presubmit and run smoketest conditionally.

### PR DESCRIPTION
**Pull request by ethanbao**
The golden test (including the new docker image building using cache) took about 9 mins (e.g. https://circleci.com/gh/ethanbao/artman/2960)

Running golden test in this way will help capture artman output change during the PR presubmit.

### CODE REVIEW

File: .circleci/config.yml [hunk 1]
```diff
@@ -45,10 +45,15 @@ jobs:
       - run:
           name: Run smoketests
           command: |
-            docker run -it -e RUNNING_IN_ARTMAN_DOCKER=True --name smoketest artman /bin/bash -c "git clone https://github.com/googleapis/googleapis /tmp/googleapis_smoketest; python3 /artman/test/smoketest_artman.py --root-dir=/tmp/googleapis_smoketest --log=/tmp/smoketest.log" || OUT=$?
-            docker cp smoketest:/tmp/smoketest.log /tmp/smoketest.log
-            if [ $OUT -ne 0 ];then
-              exit $OUT
+            if [ "${RUN_SMOKETEST}"]; then
```
Comment (geigerj) [inline .circleci/config.yml]: nit: space before `]`

File: .circleci/config.yml [hunk 2]
```diff
@@ -316,6 +321,8 @@ workflows:
       - smoke-php
       - smoke-python
       - smoke-ruby
+      # TODO(ethanbao): Separate the test and docker release into two jobs.
+      - build_and_push_docker_image
```
Comment (geigerj) [inline .circleci/config.yml]: Why is this being done twice? Also, without a branch filter like below, does this mean that docker images might get pushed from arbitrary branches?

Comment (ethanbao) [reply .circleci/config.yml]: Good catch. The docker push should only happen for master branch.

**Comment by ethanbao** [comment]
For googleapis/artman CircleCI project, I have set this `RUN_SMOKETEST` env variable. For the forked repos, developers can opt to set the env variable as well, but that will slow down the tests.

The purpose of this change is to add golden test back (under `build_and_push_docker` name though). The way we used to smoketest artman is problematic because it doesn't rebuild the Docker (therefore won't pick up the latest shared config change, which will affect the artman output). What happened is the PR gets merged with all tests passing, while one future PR will receive a confusing golden test failure report which is due to an earlier PR.

**Comment by geigerj** [comment]
> For the forked repos, developers can opt to set the env variable as well, but that will slow down the tests.

I wouldn't worry about this. I doubt most forks care about testing or accept PRs.

File: .circleci/config.yml [hunk 3]
```diff
@@ -45,11 +45,55 @@ jobs:
       - run:
           name: Run smoketests
           command: |
-            docker run -it -e RUNNING_IN_ARTMAN_DOCKER=True --name smoketest artman /bin/bash -c "git clone https://github.com/googleapis/googleapis /tmp/googleapis_smoketest; python3 /artman/test/smoketest_artman.py --root-dir=/tmp/googleapis_smoketest --log=/tmp/smoketest.log" || OUT=$?
-            docker cp smoketest:/tmp/smoketest.log /tmp/smoketest.log
-            if [ $OUT -ne 0 ];then
-              exit $OUT
+            if [ "${RUN_SMOKETEST}" ]; then
```
Comment (geigerj) [inline .circleci/config.yml]: I still don't know why this needs to be flag-guarded, but won't block the review over it.

File: .circleci/config.yml [hunk 4]
```diff
@@ -45,11 +45,55 @@ jobs:
...
+
+  build_and_release_docker_image:
+    working_directory: /usr/src/artman
+    docker:
+      - image: docker:17.05.0-ce-git
+    steps:
+      - checkout
+      - setup_remote_docker
+      - run:
+          name: Fingers crossed!
```
Comment (geigerj) [inline .circleci/config.yml]: Why do we have these `echo` commands?

File: .circleci/config.yml [hunk 5]
```diff
@@ -45,11 +45,55 @@ jobs:
...
+          command: |
+            echo "Fingers crossed!"
+      - restore_cache:
```
Comment (geigerj) [inline .circleci/config.yml]: Is there a way to factor out this set-up? If unknown, a TODO?

File: .circleci/config.yml (hunk 3 above)
Comment (ethanbao) [reply .circleci/config.yml]: The whole smoketest took ~1 hour, I don't want to have it running during each PR review, but want to have it running automatically when the PR gets merged. I didn't figure out other way except using the circleci project env variable.

File: .circleci/config.yml (hunk 4 above)
Comment (ethanbao) [reply .circleci/config.yml]: For fun

File: .circleci/config.yml (hunk 3 above)
Comment (geigerj) [reply .circleci/config.yml]: >  I don't want to have it running during each PR review, but want to have it running automatically when the PR gets merged.

Doesn't a branch filter on `master` work for this?

File: .circleci/config.yml (hunk 5 above)
Comment (ethanbao) [reply .circleci/config.yml]: `restore_cache` is a circleci built-in task. What do you mean by factor out?

File: .circleci/config.yml (hunk 3 above)
Comment (geigerj) [reply .circleci/config.yml]: FWIW, the way acceptance tests are set up in Ruby is that they run only if `CIRCLE_BRANCH` == `master` and `CIRCLE_PR_NUMBER` is empty.

File: .circleci/config.yml (hunk 5 above)
Comment (geigerj) [reply .circleci/config.yml]: I mean that lines 74-96 duplicate lines 15-37.

File: .circleci/config.yml (hunk 3 above)
Comment (ethanbao) [reply .circleci/config.yml]: Probably will work in practice, as the PR is most likely from a non-master forked branch. But PR like this (https://github.com/googleapis/artman/pull/386) will trigger the full smoketest.

**Comment by ethanbao** [comment]
PTAL @geigerj

**Pull request merged**